drivers/crypto/crypto_mtls_shim.c

```diff
@@ -99,37 +142,48 @@ static int mtls_ccm_decrypt_auth(struct cipher_ctx *ctx,
 	return 0;
 }
 
-static int mtls_get_unused_session_index(void)
+static struct mtls_shim_session *mtls_alloc_session(enum cipher_algo algo)
 {
-	int i;
+	struct mtls_shim_session *ret;
 
-	for (i = 0; i < CRYPTO_MAX_SESSION; i++) {
-		if (!mtls_sessions[i].in_use) {
-			mtls_sessions[i].in_use = true;
-			return i;
-		}
+	ret = k_mem_pool_malloc(&mtls_shim_session_pool, sizeof(*ret));
+	if (ret == NULL) {
+		SYS_LOG_ERR("No free session for now");
+	} else {
+		ret->algo = algo;
 	}
 
-	return -1;
+	return ret;
 }
 
-static int mtls_session_setup(struct device *dev, struct cipher_ctx *ctx,
-		       enum cipher_algo algo, enum cipher_mode mode,
-		       enum cipher_op op_type)
+static void mtls_free_session(struct mtls_shim_session *session)
 {
-	mbedtls_ccm_context *mtls_ctx;
-	int ctx_idx;
-	int ret;
-
-	if (ctx->flags & ~(MTLS_SUPPORT)) {
-		SYS_LOG_ERR("Unsupported flag");
-		return -EINVAL;
+	switch (session->algo) {
+	case CRYPTO_CIPHER_ALGO_AES:
+		mbedtls_ccm_free(&session->mtls);
+		break;
+	case CRYPTO_CIPHER_ALGO_PRNG:
+		mbedtls_hmac_drbg_free(&session->hmac_drbg);
+		break;
+	case CRYPTO_CIPHER_ALGO_MAC:
+		mbedtls_cipher_free(&session->cipher);
+		break;
+	case CRYPTO_CIPHER_ALGO_ECC:
+		/* Do nothing. */
+		break;
 	}
 
-	if (algo != CRYPTO_CIPHER_ALGO_AES) {
-		SYS_LOG_ERR("Unsupported algo");
-		return -EINVAL;
-	}
+	memset(session, 0, sizeof(*session));
```

Comment:
isn't the case of using `explicit_bzero()` ?

Reply:
Doesn't hurt to use it here, yes.